enhancements/cluster-capabilities-api.md

```diff
@@ -0,0 +1,328 @@
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+As we add more deployment topologies for OpenShift, we need a way to
+communicate expectations to the operators managing the cluster. This
+enhancement describes an API for expressing the expected
+"capabilities" of a cluster, including aspects such as how hard
+operators should work to provide high-availability and whether
+operators should configure their operands in a way to minimize
+resource consumption for especially constrained environments.
+
+## Motivation
+
+We want to avoid having cluster profiles use different manifests to
+configure operators managed by the cluster-version-operator, and
+instead rely on in-cluster APIs. This approach will make the same
+information available to operators managed by either the
+cluster-version-operator or operator-lifecycle-manager.
+
+We do not want to expose the cluster profile name directly to most of
+the operators, because we do not want to have to update operators each
+time we add a new profile.
+
+We want the settings to be global to the cluster, rather than using
+operator-specific APIs, because that simplifies the integration with
+the installer and allows them to be applied consistently in all
+operators that need them.
+
+### Goals
+
+- Expose capability expectations to operators running in a cluster so
+  they can configure their operands appropriately.
+
+### Non-Goals
+
+- Expose cluster profile names to operators.
+- Provide a mutable API for changing the cluster capabilities (at
+  least for now).
+- Predict other capability settings that we might need to expose.
+
+## Proposal
+
+A new [informational
+resource](https://docs.openshift.com/container-platform/4.6/installing/install_config/customizations.html#informational-resources_customizations),
+`capabilities.config.openshift.io`, is created to hold details about
+the configured features of a cluster. It will include two fields to
+start, and may be expanded over time.
+
+The `highAvailabilityMode` field will use an enum to express the
+high-availability expectations. Possible values are `full` and
+`none`. The default is `full`, which represents the behavior operators
+have today in a "normal" cluster. The `none` mode will be used in
+single-node deployments (developer and production), and the operator
+should not configure the operand for highly-available
+operation. Systems such as the assisted installer, which may
+experience periods early in the lifetime of the cluster when full HA
+is not possible, can start with `none` and transition to `full` when
+the cluster is more fully online.
+
+The `resourceLimits` fields will use an enum to express the general
```

Comment:
So are we not breaking this down by resource type ([`cpu`, `memory`, etc.][1])?  On the one hand, it's not clear to me how many components can scale their consumption of different resource types independently (e.g. reducing memory use while leaving CPU consumption high).  On the other hand, it would be a shame to scale down a CPU-intensive but memory/disk-light component if you had plenty of CPU but limited amounts of memory.  It may be worth going through the heavy components (monitoring, etc.) and providing more details about their current consumption and the expected reductions needed to fit into some hypothetical target system.

[1]: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/

Reply:
Yeah, we did talk about using separate fields for different resource types, but it seemed better to keep it simple.

Reply:
I like the intent of this but it feels like it's going to be subject to a lot more debate - maybe this one is a separate enhancement?  I'd like to get the centralized field for single node first personally.